Remove the tree viewset

rtibbles opened the pull request:
## Description

* The relatively straight forward proposition of this PR is to remove the tree viewset and just go back to querying contentnodes
* This is not so straight forward as much of the frontend and backend code is built predicated on this separation
* Refactor all frontend ContentNode vuex state to just work directly with ContentNodes and not separate tree node representations
* Update all copy operations for ContentNodes in the tree to work on just a ContentNode
* Update resource loading on the channel edit page
* Switch caching strategy from matching indexedDB queries to parameter specific in order to prevent synchronous requests from hitting the backend twice.
* Update the clipboard to no longer rely on the Tree
* Create a separate Clipboard viewset to handle incremental loading of the clipboard and node creation within the clipboard
* Update the clipboard copy strategy to only copy children when they differ from the original node children
* Dynamically load children in the clipboard depending on whether it is a deep representation
* Add logging messages in the resource layer to allow for easier stack traces to understand where async calls originate
* Update the sync code to sync more frequently


#### Issue Addressed (if applicable)

Fixes #2105
Fixes #1941
Fixes #1882


## Steps to Test

* Check that channels with large trees load channel edit
* Check that you can copy content nodes (caveat - assessmentitem file copying is currently broken from a pre-existing issue)
* Check that you can copy content nodes to and from the clipboard

nucleogenesis commented on contentcuration/contentcuration/frontend/channelEdit/views/CurrentTopicView.vue:
```diff
@@ -379,15 +373,17 @@
       }),
-      copyToClipboard: withChangeTracker(function(id__in, changeTracker) {
+      copyToClipboard: withChangeTracker(function(ids, changeTracker) {
+        const nodes = this.getContentNodes(ids);
```
This getter isn't defined here yet:

`vue.runtime.esm.js:619 [Vue warn]: Error in v-on handler (Promise/async): "TypeError: this.getContentNodes is not a function"`

nucleogenesis replied:
I found all of this testing the clipboard. After I added `getContentNodes` to the file locally to see if it was the only thing stopping clipboard from working, I get this error:

`vue.runtime.esm.js:619 [Vue warn]: Error in v-on handler (Promise/async): "DataError: Failed to execute 'only' on 'IDBKeyRange': The parameter is not a valid key."`

rtibbles replied:
That error will happen when `undefined` gets passed in as a key parameter to IndexedDB. Guessing the getContentNodes got lost in the rebase. Thanks for testing here.

rtibbles replied:
Managed to replicate and resolve!

nucleogenesis commented:
Possibly unrelated: Removing a Topic (only topics as far as I can see) from a channel looks like it works at first (snackbar is good, node removed from node panel) but on page reload after a few seconds, it reappears. Content items are moved to trash as expected.

nucleogenesis commented:
Great work @rtibbles this is a lot. The JS/Vue looked good to me. I noted a couple issues above and will be happy to test again whenever you'd like another pass.

rtibbles commented:
Have resolved test failures and issues observed from manual testing, think this is ready for broader review now.

rtibbles commented:
One last linting fix.

rtibbles commented:
OK, for real, that's the last linting fix...

nucleogenesis reviewed (approved):
@rtibbles Fixes work for me and lent to me finding some new fun bugs in the clipboard (not related to this PR)!

bjester commented on contentcuration/contentcuration/frontend/shared/data/changes.js:
```diff
@@ -172,7 +172,7 @@ export class ChangeTracker extends EventEmitter {
-
+    this._timeout = setTimeout(this.dismiss.bind(this), this.expiry);
     // This sets the expiry, from 0 to an actual expiry. The clock is ticking...
     return this.renew(this.expiry);
```
Not a big deal, it may make sense to have the timer set within the `renew` method.

bjester commented on contentcuration/contentcuration/frontend/shared/data/resources.js:
```diff
@@ -437,10 +436,22 @@ class Resource extends mix(APIResource, IndexedDBResource) {
+      this._requests[queryString][LAST_FETCHED] + REFRESH_INTERVAL * 1000 > now &&
+      this._requests[queryString].promise
+    ) {
+      return this._requests[queryString].promise;
```
Perhaps assign `this._requests[queryString]` to a variable to save some bytes in minification.

rtibbles replied:
Yeah, wfm.

bjester reviewed (approved):
Aww I kind of miss the tree data structure... Oh well. Looks good to me.

nucleogenesis reviewed (changes requested):
I've found some more issues that may be related to the PR.

In the following (some may not be related to PR):

- Topic says 6 items, opens to show none.
- Content items are indented related to topics - looking like children of a node they are not children of.
- The only time the context menu options appear at the top of the clipboard drawer is when a whole channel's clipboard node is selected.
- In NodePanel, selecting more than one item and copying it doesn't copy everything.

![weird-clipboard](https://user-images.githubusercontent.com/6356129/92291033-7f50e080-eecb-11ea-952f-a6a2eee0bdc0.gif)

This happened once:

- Have nodes copied to clipboard.
- Delete those nodes
- See the actual nodes in the channel get deleted... :( 

I did this once and the nodes were deleted in NodePanel and Clipboard but when I refreshed they came back. So I decided to record doing it again for this and when the nodes got deleted in both places it was permanent.


![toomanydeletions](https://user-images.githubusercontent.com/6356129/92290855-98a55d00-eeca-11ea-9687-2746be95318a.gif)

bjester commented:
@nucleogenesis Ah yeah, doesn't look like the clipboard selection handling is working correctly, at the least. It should look like this when a deep node is selected. Admittedly, I know I left 1 or 2 bugs with the selection logic, but they're not obvious. 

![Screenshot from 2020-09-04 16-39-32](https://user-images.githubusercontent.com/819838/92291454-9a245480-eecd-11ea-90f1-b5b7364f717e.png)

@rtibbles Let me know if you want any eyes on the selection logic.

rtibbles commented:
I have fixed the deletion issues - there is one extant thing that I am aware of to do with descendant counts in the clipboard.

Also, I noticed that toggling a topic's selection toggles through:
* Topic selected and all descendants selected
* Topic indeterminate and all descendants selected
* Topic selected and all descendants unselected
* Topic unselected and all descendants unselected

I tried to fix this, but the bit maps made my brain hurt. As it is pre-existing, I'd rather leave it.

rtibbles commented:
Going to rebase to address merge conflicts.

rtibbles commented:
Rebased.

rtibbles commented:
@nucleogenesis I think I have addressed most of your issues, but
> Topic says 6 items, opens to show none.
May still be extant in places, depending on the exact sequence of actions you take. Would like to come back to that later as part of some other work.

nucleogenesis commented:
1.GET http://localhost:8080/api/contentnode?parent=53ffb3c1131c47dbaa11ddafcf01710a -> `412 Precondition Failed` -> `{"detail":"No valid filter parameters supplied"}`

I 'duplicated' a topic folder that had some contents in it. When I navigated to that copy, I get the above error. Note that this is unrelated to the clipboard itself, but perhaps shares some internals?

~2. One of the pre-existing clipboard nodes no longer appears to exist. I cannot select it in the Clipboard, which appears to prevent my selecting the entire channel that the bad node lives in. ~

3. Whenever I try to open the clipboard after adding nodes, I get this error. It seems that I've now gotten my clipboard into an unrecoverable broken state by having added some nodes to it.

```
TypeError: Cannot read property 'excluded_descendants' of null
    at getters.js:77
    at Array.filter (<anonymous>)
    at Object.<anonymous> (getters.js:77)
    at Object.<anonymous> (getters.js:303)
    at getters.js:354
    at Array.map (<anonymous>)
    at Object.<anonymous> (getters.js:354)
    at VueComponent.<anonymous> (getters.js:312)
    at VueComponent.selectionState (mixins.js:33)
    at Watcher.get (vue.runtime.esm.js:4479)
```

Just need to handle the case where the `excluded_descendants` isn't yet set. I fixed it locally using `lodash/get` and appears to work now. Nodes that I tried to copy before it was "broken" that never showed are now showing. This also fixes #2 above as I was able to select that node.

```
import get from 'lodash/get';

...
if(ancestor) {
    const filterFn = c => get(ancestor, 'extra_fields.excluded_descendants[c.id]')  
    return children.filter(filterFn); 
}
```

4. I moved a video from root tree node to a newly created subtopic, but the move is not reflected in either place. I also got a 400 on the sync endpoint relatively close to trying this and prettified the payload and response here: https://gist.github.com/nucleogenesis/21769a23a812ce5c8ff17b9d9e85d18e 

5. I copied from Channel A to Clipboard and when I went to insert into Channel B (an entirely empty channel) it did not show in the NodePanel. However, when I go to move another node from Clipboard -> Channel B again, the first file I moved there is showing in Channel B's root on the "Select where you want to move this clipboard node" page. On the Channel List page, I see the proper resource count (I copied two files there) but they are not showing in the NodePanel after several refreshes and navigating away and back.

Seems only to be an issue of putting nodes in the channel root when it is empty. Otherwise, it appears to work as expected.

Also note that the last 2 items were tested with my fix for the `excluded_descendants` in place - so if your solution differs fundamentally, then I'll revisit these once fixed.

nucleogenesis reviewed (approved):
LGTM!
